File: android/app/src/main/java/com/kaspriv/wallet/HardwareVaultPlugin.java
package com.kaspriv.wallet;

import android.os.Build;
import android.security.keystore.KeyGenParameterSpec;
import android.security.keystore.KeyInfo;
import android.security.keystore.KeyPermanentlyInvalidatedException;
import android.security.keystore.KeyProperties;
import android.util.Base64;

import androidx.annotation.NonNull;
import androidx.biometric.BiometricManager;
import androidx.biometric.BiometricPrompt;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentActivity;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.security.KeyStore;
import java.util.Arrays;
import java.util.concurrent.Executor;
import java.util.concurrent.atomic.AtomicBoolean;

import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.GCMParameterSpec;

@CapacitorPlugin(name = "HardwareVault")
public class HardwareVaultPlugin extends Plugin {

    private static final String ANDROID_KEYSTORE = "AndroidKeyStore";
    private static final String TRANSFORMATION = "AES/GCM/NoPadding";

    private static final int GCM_TAG_BITS = 128;
    private static final int GCM_IV_BYTES = 12;
    private static final int KEY_SIZE = 256;

    // Prevent unexpectedly large encrypted vault payloads.
    private static final int MAX_CIPHERTEXT_BYTES = 4096;

    // Native namespace for all KasPriv biometric keys.
    private static final String ALIAS_PREFIX = "kaspriv_vault_";

    /**
     * The decrypted secret execution handler.
     * Consumes secret synchronously in memory and returns resulting non-secret payload or secret Base64 for vault decryption.
     */
    private interface NativeSecretOperation {
        JSObject execute(byte[] secret, PluginCall call) throws Exception;
    }

    @PluginMethod
    public void checkBiometricAvailability(PluginCall call) {
        try {
            BiometricManager biometricManager =
                    BiometricManager.from(getContext());

            int status = biometricManager.canAuthenticate(
                    BiometricManager.Authenticators.BIOMETRIC_STRONG
            );

            String reason;

            switch (status) {
                case BiometricManager.BIOMETRIC_SUCCESS:
                    reason = "available";
                    break;

                case BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE:
                    reason = "no_hardware";
                    break;

                case BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE:
                    reason = "hw_unavailable";
                    break;

                case BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED:
                    reason = "not_enrolled";
                    break;

                case BiometricManager.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED:
                    reason = "security_update_required";
                    break;

                default:
                    reason = "unknown";
                    break;
            }

            JSObject result = new JSObject();
            result.put(
                    "available",
                    status == BiometricManager.BIOMETRIC_SUCCESS
            );
            result.put("reason", reason);

            call.resolve(result);

        } catch (Exception e) {
            call.reject("BIOMETRIC_CHECK_FAILED");
        }
    }

    @PluginMethod
    public void createBiometricKey(PluginCall call) {

        String alias = call.getString(
                "alias",
                ALIAS_PREFIX + "v1"
        );

        boolean requireStrongBox = Boolean.TRUE.equals(call.getBoolean("requireStrongBox", false));

        if (!isValidAlias(alias)) {
            call.reject("ALIAS_OUT_OF_NAMESPACE");
            return;
        }

        try {
            KeyStore keyStore =
                    KeyStore.getInstance(ANDROID_KEYSTORE);

            keyStore.load(null);

            if (keyStore.containsAlias(alias)) {
                SecretKey key = (SecretKey) keyStore.getKey(alias, null);
                JSObject infoObj = inspectKeyInfo(key);

                JSObject result = new JSObject();
                result.put("alias", alias);
                result.put("existed", true);
                result.put("isHardwareBacked", infoObj.getBoolean("isHardwareBacked"));
                result.put("securityLevel", infoObj.getString("securityLevel"));

                call.resolve(result);
                return;
            }

            if (requireStrongBox && Build.VERSION.SDK_INT < Build.VERSION_CODES.P) {
                call.reject("STRONGBOX_UNAVAILABLE");
                return;
            }

            KeyGenerator keyGenerator =
                    KeyGenerator.getInstance(
                            KeyProperties.KEY_ALGORITHM_AES,
                            ANDROID_KEYSTORE
                    );

            KeyGenParameterSpec.Builder builder =
                    new KeyGenParameterSpec.Builder(
                            alias,
                            KeyProperties.PURPOSE_ENCRYPT
                                    | KeyProperties.PURPOSE_DECRYPT
                    )
                            .setBlockModes(
                                    KeyProperties.BLOCK_MODE_GCM
                            )
                            .setEncryptionPaddings(
                                    KeyProperties.ENCRYPTION_PADDING_NONE
                            )
                            .setKeySize(KEY_SIZE)
                            .setUserAuthenticationRequired(true)
                            .setInvalidatedByBiometricEnrollment(true);

            if (requireStrongBox && Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                try {
                    builder.setIsStrongBoxBacked(true);
                } catch (Exception e) {
                    call.reject("STRONGBOX_UNAVAILABLE");
                    return;
                }
            }

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                builder.setUserAuthenticationParameters(
                        0,
                        KeyProperties.AUTH_BIOMETRIC_STRONG
                );
            } else {
                builder.setUserAuthenticationValidityDurationSeconds(-1);
            }

            try {
                keyGenerator.init(builder.build());
                keyGenerator.generateKey();
            } catch (Exception e) {
                if (requireStrongBox) {
                    call.reject("STRONGBOX_UNAVAILABLE");
                    return;
                } else {
                    throw e;
                }
            }

            SecretKey createdKey = (SecretKey) keyStore.getKey(alias, null);
            JSObject infoObj = inspectKeyInfo(createdKey);

            JSObject result = new JSObject();
            result.put("alias", alias);
            result.put("existed", false);
            result.put("isHardwareBacked", infoObj.getBoolean("isHardwareBacked"));
            result.put("securityLevel", infoObj.getString("securityLevel"));

            call.resolve(result);

        } catch (Exception e) {
            call.reject("CREATE_KEY_FAILED");
        }
    }

    @PluginMethod
    public void wrapSecret(PluginCall call) {
        String alias = call.getString("alias");
        String secretBase64 = call.getString("secretBase64");

        if (!isValidAlias(alias)) {
            call.reject("ALIAS_OUT_OF_NAMESPACE");
            return;
        }

        if (secretBase64 == null || secretBase64.isEmpty()) {
            call.reject("MISSING_SECRET");
            return;
        }

        FragmentActivity activity = getActivity();
        if (activity == null) {
            call.reject("NO_ACTIVITY");
            return;
        }

        byte[] secret = null;
        boolean wrappedOwnershipTransferred = false;

        try {
            try {
                secret = Base64.decode(secretBase64, Base64.NO_WRAP);
            } catch (IllegalArgumentException e) {
                call.reject("INVALID_BASE64");
                return;
            }

            KeyStore keyStore = KeyStore.getInstance(ANDROID_KEYSTORE);
            keyStore.load(null);

            SecretKey key = (SecretKey) keyStore.getKey(alias, null);
            if (key == null) {
                call.reject("KEY_NOT_FOUND");
                return;
            }

            Cipher cipher = Cipher.getInstance(TRANSFORMATION);
            cipher.init(Cipher.ENCRYPT_MODE, key);

            BiometricPrompt.CryptoObject cryptoObject =
                    new BiometricPrompt.CryptoObject(cipher);

            NativeSecretOperation operation =
                    (ciphertext, nativeCall) -> {
                        byte[] iv = cipher.getIV();
                        JSObject ret = new JSObject();
                        ret.put("wrappedBase64", Base64.encodeToString(ciphertext, Base64.NO_WRAP));
                        ret.put("ivBase64", Base64.encodeToString(iv, Base64.NO_WRAP));
                        return ret;
                    };

            prompt(
                    activity,
                    cryptoObject,
                    "Register KasPriv vault",
                    call,
                    secret,
                    operation
            );

            wrappedOwnershipTransferred = true;

        } catch (Exception e) {
            call.reject("WRAP_SECRET_FAILED");
        } finally {
            if (!wrappedOwnershipTransferred) {
                wipe(secret);
            }
        }
    }

    @PluginMethod
    public void hasKey(PluginCall call) {
        String alias = call.getString("alias");
        if (!isValidAlias(alias)) {
            call.reject("ALIAS_OUT_OF_NAMESPACE");
            return;
        }
        try {
            KeyStore keyStore = KeyStore.getInstance(ANDROID_KEYSTORE);
            keyStore.load(null);
            boolean exists = keyStore.containsAlias(alias);
            JSObject result = new JSObject();
            result.put("exists", exists);
            call.resolve(result);
        } catch (Exception e) {
            call.reject("HAS_KEY_FAILED");
        }
    }

    @PluginMethod
    public void deleteKey(PluginCall call) {
        String alias = call.getString("alias");
        if (!isValidAlias(alias)) {
            call.reject("ALIAS_OUT_OF_NAMESPACE");
            return;
        }
        deleteKeystoreEntry(alias);
        call.resolve();
    }

    @PluginMethod
    public void unwrapSecret(PluginCall call) {

        String alias = call.getString("alias");
        String wrappedBase64 = call.getString("wrappedBase64");
        String ivBase64 = call.getString("ivBase64");

        if (!isValidAlias(alias)) {
            call.reject("ALIAS_OUT_OF_NAMESPACE");
            return;
        }

        if (wrappedBase64 == null || wrappedBase64.isEmpty()) {
            call.reject("MISSING_CIPHERTEXT");
            return;
        }

        if (ivBase64 == null || ivBase64.isEmpty()) {
            call.reject("MISSING_IV");
            return;
        }

        FragmentActivity activity = getActivity();

        if (activity == null) {
            call.reject("NO_ACTIVITY");
            return;
        }

        byte[] wrapped = null;
        byte[] iv = null;

        boolean wrappedOwnershipTransferred = false;

        try {

            try {
                wrapped = Base64.decode(
                        wrappedBase64,
                        Base64.NO_WRAP
                );

                iv = Base64.decode(
                        ivBase64,
                        Base64.NO_WRAP
                );

            } catch (IllegalArgumentException e) {
                call.reject("INVALID_BASE64");
                return;
            }

            if (iv.length != GCM_IV_BYTES) {
                call.reject("INVALID_IV_LENGTH");
                return;
            }

            if (wrapped.length <= GCM_TAG_BITS / 8) {
                call.reject("INVALID_CIPHERTEXT_LENGTH");
                return;
            }

            if (wrapped.length > MAX_CIPHERTEXT_BYTES) {
                call.reject("INVALID_CIPHERTEXT_LENGTH");
                return;
            }

            KeyStore keyStore =
                    KeyStore.getInstance(ANDROID_KEYSTORE);

            keyStore.load(null);

            SecretKey key =
                    (SecretKey) keyStore.getKey(alias, null);

            if (key == null) {
                call.reject("KEY_NOT_FOUND");
                return;
            }

            Cipher cipher =
                    Cipher.getInstance(TRANSFORMATION);

            cipher.init(
                    Cipher.DECRYPT_MODE,
                    key,
                    new GCMParameterSpec(
                            GCM_TAG_BITS,
                            iv
                    )
            );

            BiometricPrompt.CryptoObject cryptoObject =
                    new BiometricPrompt.CryptoObject(cipher);

            NativeSecretOperation operation =
                    (secret, nativeCall) -> {
                        JSObject ret = new JSObject();
                        ret.put("secretBase64", Base64.encodeToString(secret, Base64.NO_WRAP));
                        return ret;
                    };

            prompt(
                    activity,
                    cryptoObject,
                    "Unlock KasPriv vault",
                    call,
                    wrapped,
                    operation
            );

            wrappedOwnershipTransferred = true;

        } catch (KeyPermanentlyInvalidatedException e) {

            deleteKeystoreEntry(alias);

            call.reject(
                    "KEY_INVALIDATED_REENROLL_REQUIRED"
            );

        } catch (Exception e) {

            call.reject("UNWRAP_INIT_FAILED");

        } finally {

            wipe(iv);

            if (!wrappedOwnershipTransferred) {
                wipe(wrapped);
            }
        }
    }

    private void prompt(
            FragmentActivity activity,
            BiometricPrompt.CryptoObject cryptoObject,
            String title,
            PluginCall call,
            byte[] wrapped,
            NativeSecretOperation operation
    ) {

        Executor executor =
                ContextCompat.getMainExecutor(activity);

        AtomicBoolean settled =
                new AtomicBoolean(false);

        BiometricPrompt.PromptInfo promptInfo =
                new BiometricPrompt.PromptInfo.Builder()
                        .setTitle(title)
                        .setSubtitle("Biometric authentication")
                        .setNegativeButtonText("Cancel")
                        .setAllowedAuthenticators(
                                BiometricManager.Authenticators.BIOMETRIC_STRONG
                        )
                        .build();

        activity.runOnUiThread(() -> {

            BiometricPrompt biometricPrompt =
                    new BiometricPrompt(
                            activity,
                            executor,
                            new BiometricPrompt.AuthenticationCallback() {

                                @Override
                                public void onAuthenticationSucceeded(
                                        @NonNull BiometricPrompt.AuthenticationResult result
                                ) {

                                    if (!settled.compareAndSet(
                                            false,
                                            true
                                    )) {
                                        return;
                                    }

                                    byte[] secret = null;

                                    try {

                                        Cipher cipher =
                                                result.getCryptoObject()
                                                        .getCipher();

                                        if (cipher == null) {
                                            throw new IllegalStateException(
                                                    "NO_CIPHER"
                                            );
                                        }

                                        secret = cipher.doFinal(wrapped);

                                        JSObject resultObject =
                                                operation.execute(
                                                        secret,
                                                        call
                                                );

                                        call.resolve(resultObject);

                                    } catch (UnsupportedOperationException e) {

                                        call.reject(
                                                "NATIVE_SIGNER_NOT_CONFIGURED"
                                        );

                                    } catch (javax.crypto.AEADBadTagException e) {

                                        call.reject(
                                                "DECRYPT_AUTH_FAILED"
                                        );

                                    } catch (Exception e) {

                                        call.reject(
                                                "BIOMETRIC_OPERATION_FAILED"
                                        );

                                    } finally {

                                        wipe(secret);
                                        wipe(wrapped);
                                    }
                                }

                                @Override
                                public void onAuthenticationError(
                                        int errorCode,
                                        @NonNull CharSequence errString
                                ) {

                                    if (!settled.compareAndSet(
                                            false,
                                            true
                                    )) {
                                        return;
                                    }

                                    wipe(wrapped);

                                    String reason;

                                    switch (errorCode) {

                                        case BiometricPrompt.ERROR_LOCKOUT:
                                        case BiometricPrompt.ERROR_LOCKOUT_PERMANENT:
                                            reason = "lockout";
                                            break;

                                        case BiometricPrompt.ERROR_HW_UNAVAILABLE:
                                            reason = "hw_unavailable";
                                            break;

                                        case BiometricPrompt.ERROR_NO_BIOMETRICS:
                                            reason = "not_enrolled";
                                            break;

                                        case BiometricPrompt.ERROR_USER_CANCELED:
                                        case BiometricPrompt.ERROR_NEGATIVE_BUTTON:
                                            reason = "user_cancelled";
                                            break;

                                        default:
                                            reason = "other";
                                            break;
                                    }

                                    JSObject error =
                                            new JSObject();

                                    error.put(
                                            "errorCode",
                                            errorCode
                                    );

                                    error.put(
                                            "reason",
                                            reason
                                    );

                                    call.reject(
                                            reason,
                                            error
                                    );
                                }

                                @Override
                                public void onAuthenticationFailed() {
                                }
                            }
                    );

            try {

                biometricPrompt.authenticate(
                        promptInfo,
                        cryptoObject
                );

            } catch (Exception e) {

                if (settled.compareAndSet(false, true)) {

                    wipe(wrapped);

                    call.reject(
                            "BIOMETRIC_PROMPT_FAILED"
                    );
                }
            }
        });
    }

    private static JSObject inspectKeyInfo(SecretKey key) {
        boolean isHardwareBacked = false;
        String securityLevel = "unknown";
        if (key != null) {
            try {
                SecretKeyFactory factory = SecretKeyFactory.getInstance(key.getAlgorithm(), ANDROID_KEYSTORE);
                KeyInfo info = (KeyInfo) factory.getKeySpec(key, KeyInfo.class);
                isHardwareBacked = info.isInsideSecureHardware();
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                    int level = info.getSecurityLevel();
                    switch (level) {
                        case KeyProperties.SECURITY_LEVEL_STRONGBOX:
                            securityLevel = "strongbox";
                            break;
                        case KeyProperties.SECURITY_LEVEL_TRUSTED_ENVIRONMENT:
                            securityLevel = "tee";
                            break;
                        case KeyProperties.SECURITY_LEVEL_SOFTWARE:
                            securityLevel = "software";
                            break;
                        default:
                            securityLevel = isHardwareBacked ? "hardware" : "software";
                            break;
                    }
                } else {
                    securityLevel = isHardwareBacked ? "tee" : "software";
                }
            } catch (Exception ignored) {}
        }
        JSObject result = new JSObject();
        result.put("isHardwareBacked", isHardwareBacked);
        result.put("securityLevel", securityLevel);
        return result;
    }

    private static boolean isValidAlias(String alias) {

        if (alias == null) {
            return false;
        }

        if (alias.isEmpty()) {
            return false;
        }

        if (!alias.startsWith(ALIAS_PREFIX)) {
            return false;
        }

        if (alias.length() > 128) {
            return false;
        }

        return true;
    }

    private static void deleteKeystoreEntry(String alias) {

        if (alias == null) {
            return;
        }

        try {

            KeyStore keyStore =
                    KeyStore.getInstance(
                            ANDROID_KEYSTORE
                    );

            keyStore.load(null);

            if (keyStore.containsAlias(alias)) {
                keyStore.deleteEntry(alias);
            }

        } catch (Exception ignored) {
        }
    }

    private static void wipe(byte[] data) {

        if (data != null) {
            Arrays.fill(data, (byte) 0);
        }
    }
}

File: android/app/src/main/java/com/kaspriv/wallet/HardwareVaultPlugin.kt
package com.kaspriv.wallet

import android.os.Build
import android.os.Handler
import android.os.Looper
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyInfo
import android.security.keystore.KeyProperties
import android.util.Base64
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import java.security.KeyStore
import java.util.concurrent.atomic.AtomicBoolean
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

@CapacitorPlugin(name = "HardwareVault")
class HardwareVaultPlugin : Plugin() {

    private val KEY_STORE_NAME = "AndroidKeyStore"
    private val TRANSFORMATION = "AES/GCM/NoPadding"
    private val TAG_LENGTH = 128
    private val mainHandler = Handler(Looper.getMainLooper())

    @PluginMethod
    fun checkBiometricSupport(call: PluginCall) {
        val biometricManager = BiometricManager.from(context)
        val canAuthStrong = biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG)
        val canAuthWeak = biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK)
        val canAuthDevice = biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG or BiometricManager.Authenticators.DEVICE_CREDENTIAL)
        
        val isAvailable = canAuthStrong == BiometricManager.BIOMETRIC_SUCCESS || 
                          canAuthWeak == BiometricManager.BIOMETRIC_SUCCESS ||
                          canAuthDevice == BiometricManager.BIOMETRIC_SUCCESS
        
        val ret = JSObject()
        ret.put("available", isAvailable)
        ret.put("isStrong", canAuthStrong == BiometricManager.BIOMETRIC_SUCCESS)
        ret.put("status", if (canAuthStrong == BiometricManager.BIOMETRIC_SUCCESS) canAuthStrong else canAuthWeak)
        
        val hasStrongBox = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            try {
                context.packageManager.hasSystemFeature("android.hardware.strongbox_keystore")
            } catch (e: Exception) {
                false
            }
        } else {
            false
        }
        ret.put("hasStrongBox", hasStrongBox)
        
        call.resolve(ret)
    }

    @PluginMethod
    fun storeSecure(call: PluginCall) {
        val alias = call.getString("alias")
        val data = call.getString("data")
        
        if (alias == null || data == null) {
            call.reject("Alias and data are required")
            return
        }

        try {
            val secretKey = getOrCreateKey(alias)
            val cipher = Cipher.getInstance(TRANSFORMATION)
            
            try {
                cipher.init(Cipher.ENCRYPT_MODE, secretKey)
                val iv = cipher.iv
                val encryptedBytes = cipher.doFinal(data.toByteArray(Charsets.UTF_8))
                
                val ret = JSObject()
                ret.put("iv", Base64.encodeToString(iv, Base64.DEFAULT))
                ret.put("ciphertext", Base64.encodeToString(encryptedBytes, Base64.DEFAULT))
                call.resolve(ret)
            } catch (e: Exception) {
                // If auth is required for cipher operations, prompt biometrics with CryptoObject
                try {
                    cipher.init(Cipher.ENCRYPT_MODE, secretKey)
                    showBiometricPromptForEncrypt(cipher, data.toByteArray(Charsets.UTF_8), call)
                } catch (reInitErr: Exception) {
                    call.reject("Cipher initialization failed: ${reInitErr.message}")
                }
            }
        } catch (e: Exception) {
            call.reject("Failed to encrypt data: ${e.message}")
        }
    }

    @PluginMethod
    fun loadSecure(call: PluginCall) {
        val alias = call.getString("alias")
        val ivBase64 = call.getString("iv")
        val ciphertextBase64 = call.getString("ciphertext")
        
        if (alias == null || ivBase64 == null || ciphertextBase64 == null) {
            call.reject("Missing required parameters")
            return
        }

        try {
            val iv = Base64.decode(ivBase64, Base64.DEFAULT)
            val ciphertext = Base64.decode(ciphertextBase64, Base64.DEFAULT)
            
            val keyStore = KeyStore.getInstance(KEY_STORE_NAME)
            keyStore.load(null)
            val secretKey = keyStore.getKey(alias, null) as? SecretKey ?: throw Exception("Vault key not found")
            
            val cipher = Cipher.getInstance(TRANSFORMATION)
            val gcmSpec = GCMParameterSpec(TAG_LENGTH, iv)
            
            try {
                cipher.init(Cipher.DECRYPT_MODE, secretKey, gcmSpec)
                val decryptedBytes = cipher.doFinal(ciphertext)
                val ret = JSObject()
                ret.put("data", String(decryptedBytes, Charsets.UTF_8))
                wipe(decryptedBytes)
                call.resolve(ret)
            } catch (e: Exception) {
                // Decrypt requires biometric authorization
                try {
                    cipher.init(Cipher.DECRYPT_MODE, secretKey, gcmSpec)
                    showBiometricPromptForDecrypt(cipher, ciphertext, call)
                } catch (reInitErr: Exception) {
                    call.reject("Decryption cipher initialization failed: ${reInitErr.message}")
                }
            }
        } catch (e: Exception) {
            call.reject("Failed to decrypt: ${e.message}")
        }
    }

    @PluginMethod
    fun deleteKey(call: PluginCall) {
        val alias = call.getString("alias") ?: return call.reject("Alias is required")
        try {
            val keyStore = KeyStore.getInstance(KEY_STORE_NAME)
            keyStore.load(null)
            if (keyStore.containsAlias(alias)) {
                keyStore.deleteEntry(alias)
            }
            call.resolve()
        } catch (e: Exception) {
            call.reject("Failed to delete key: ${e.message}")
        }
    }

    private fun getOrCreateKey(alias: String): SecretKey {
        val keyStore = KeyStore.getInstance(KEY_STORE_NAME)
        keyStore.load(null)
        
        if (keyStore.containsAlias(alias)) {
            try {
                val existing = keyStore.getKey(alias, null) as? SecretKey
                if (existing != null) return existing
            } catch (e: Exception) {
                keyStore.deleteEntry(alias)
            }
        }

        // Check biometric enrollment to decide whether auth is required on key
        val biometricManager = BiometricManager.from(context)
        val canAuthStrong = biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG) == BiometricManager.BIOMETRIC_SUCCESS

        // 1. Try StrongBox if hardware feature is present
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            try {
                val hasStrongBox = context.packageManager.hasSystemFeature("android.hardware.strongbox_keystore")
                if (hasStrongBox) {
                    return generateAesKey(alias, useStrongBox = true, authRequired = canAuthStrong)
                }
            } catch (e: Exception) {
                // Fallback to standard hardware-backed Keystore
            }
        }

        // 2. Try standard TEE hardware-backed Keystore with biometric authentication
        try {
            return generateAesKey(alias, useStrongBox = false, authRequired = canAuthStrong)
        } catch (e: Exception) {
            // 3. Fallback without strict biometric auth flag if enrollment or hardware flags mismatch
            return generateAesKey(alias, useStrongBox = false, authRequired = false)
        }
    }

    private fun generateAesKey(alias: String, useStrongBox: Boolean, authRequired: Boolean): SecretKey {
        val keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, KEY_STORE_NAME)
        val specBuilder = KeyGenParameterSpec.Builder(
            alias,
            KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
        )
            .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
            .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
            .setKeySize(256)

        if (authRequired) {
            specBuilder.setUserAuthenticationRequired(true)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                specBuilder.setUserAuthenticationParameters(0, KeyProperties.AUTH_BIOMETRIC_STRONG)
            } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                specBuilder.setUserAuthenticationValidityDurationSeconds(-1)
            }
            // Invalidate key only if permanently removed; avoid transient disconnects
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                try {
                    specBuilder.setInvalidatedByBiometricEnrollment(false)
                } catch (e: Exception) {}
            }
        } else {
            specBuilder.setUserAuthenticationRequired(false)
        }

        if (useStrongBox && Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            try {
                specBuilder.setIsStrongBoxBacked(true)
            } catch (e: Exception) {}
        }

        keyGenerator.init(specBuilder.build())
        return keyGenerator.generateKey()
    }

    private fun showBiometricPromptForEncrypt(cipher: Cipher, plaintext: ByteArray, call: PluginCall) {
        val resolvedOrRejected = AtomicBoolean(false)
        mainHandler.post {
            try {
                val currentActivity = activity
                if (currentActivity == null || currentActivity.isFinishing || currentActivity.isDestroyed) {
                    if (resolvedOrRejected.compareAndSet(false, true)) {
                        call.reject("Activity is not available for BiometricPrompt")
                    }
                    return@post
                }

                val executor = ContextCompat.getMainExecutor(context)
                val biometricPrompt = BiometricPrompt(currentActivity, executor, object : BiometricPrompt.AuthenticationCallback() {
                    override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                        super.onAuthenticationSucceeded(result)
                        if (resolvedOrRejected.compareAndSet(false, true)) {
                            try {
                                val authCipher = result.cryptoObject?.cipher ?: cipher
                                val iv = authCipher.iv
                                val encryptedBytes = authCipher.doFinal(plaintext)
                                val ret = JSObject()
                                ret.put("iv", Base64.encodeToString(iv, Base64.DEFAULT))
                                ret.put("ciphertext", Base64.encodeToString(encryptedBytes, Base64.DEFAULT))
                                call.resolve(ret)
                            } catch (e: Exception) {
                                call.reject("Encryption failed after auth: ${e.message}")
                            }
                        }
                    }

                    override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                        super.onAuthenticationError(errorCode, errString)
                        if (resolvedOrRejected.compareAndSet(false, true)) {
                            call.reject("Authentication error ($errorCode): $errString")
                        }
                    }

                    override fun onAuthenticationFailed() {
                        super.onAuthenticationFailed()
                    }
                })

                val promptInfo = BiometricPrompt.PromptInfo.Builder()
                    .setTitle("Authenticate to Secure Wallet")
                    .setSubtitle("Confirm biometrics to save your vault key")
                    .setNegativeButtonText("Use Password")
                    .setConfirmationRequired(false)
                    .setAllowedAuthenticators(BiometricManager.Authenticators.BIOMETRIC_STRONG)
                    .build()

                biometricPrompt.authenticate(promptInfo, BiometricPrompt.CryptoObject(cipher))
            } catch (e: Exception) {
                if (resolvedOrRejected.compareAndSet(false, true)) {
                    call.reject("Biometric prompt failed: ${e.message}")
                }
            }
        }
    }

    private fun showBiometricPromptForDecrypt(cipher: Cipher, ciphertext: ByteArray, call: PluginCall) {
        val resolvedOrRejected = AtomicBoolean(false)
        mainHandler.post {
            try {
                val currentActivity = activity
                if (currentActivity == null || currentActivity.isFinishing || currentActivity.isDestroyed) {
                    if (resolvedOrRejected.compareAndSet(false, true)) {
                        call.reject("Activity is not available for BiometricPrompt")
                    }
                    return@post
                }

                val executor = ContextCompat.getMainExecutor(context)
                val biometricPrompt = BiometricPrompt(currentActivity, executor, object : BiometricPrompt.AuthenticationCallback() {
                    override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                        super.onAuthenticationSucceeded(result)
                        if (resolvedOrRejected.compareAndSet(false, true)) {
                            try {
                                val authCipher = result.cryptoObject?.cipher ?: cipher
                                val decryptedBytes = authCipher.doFinal(ciphertext)
                                val ret = JSObject()
                                ret.put("data", String(decryptedBytes, Charsets.UTF_8))
                                wipe(decryptedBytes)
                                call.resolve(ret)
                            } catch (e: Exception) {
                                call.reject("Decryption failed after auth: ${e.message}")
                            }
                        }
                    }

                    override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                        super.onAuthenticationError(errorCode, errString)
                        if (resolvedOrRejected.compareAndSet(false, true)) {
                            call.reject("Authentication error ($errorCode): $errString")
                        }
                    }

                    override fun onAuthenticationFailed() {
                        super.onAuthenticationFailed()
                    }
                })

                val promptInfo = BiometricPrompt.PromptInfo.Builder()
                    .setTitle("Unlock Kaspriv Wallet")
                    .setSubtitle("Confirm fingerprint or face biometric")
                    .setNegativeButtonText("Use Password")
                    .setConfirmationRequired(false)
                    .setAllowedAuthenticators(BiometricManager.Authenticators.BIOMETRIC_STRONG)
                    .build()

                biometricPrompt.authenticate(promptInfo, BiometricPrompt.CryptoObject(cipher))
            } catch (e: Exception) {
                if (resolvedOrRejected.compareAndSet(false, true)) {
                    call.reject("Biometric prompt failed: ${e.message}")
                }
            }
        }
    }

    private fun wipe(data: ByteArray?) {
        data?.fill(0)
    }

    @PluginMethod
    fun createBiometricKey(call: PluginCall) {
        val alias = call.getString("alias") ?: return call.reject("Alias is required")
        try {
            val keyStore = KeyStore.getInstance(KEY_STORE_NAME)
            keyStore.load(null)
            val existed = keyStore.containsAlias(alias)
            getOrCreateKey(alias)
            
            val ret = JSObject()
            ret.put("alias", alias)
            ret.put("existed", existed)
            ret.put("isHardwareBacked", true)
            ret.put("securityLevel", "hardware_keystore")
            call.resolve(ret)
        } catch (e: Exception) {
            call.reject("Failed to create biometric key: ${e.message}")
        }
    }

    @PluginMethod
    fun getHardwareSecurityLevel(call: PluginCall) {
        val alias = "security_probe_${System.currentTimeMillis()}"
        try {
            val key = getOrCreateKey(alias)
            val secretKeyFactory = javax.crypto.SecretKeyFactory.getInstance(key.algorithm, KEY_STORE_NAME)
            val keyInfo = secretKeyFactory.getKeySpec(key, KeyInfo::class.java) as KeyInfo
            
            val ret = JSObject()
            ret.put("insideSecureHardware", keyInfo.isInsideSecureHardware)
            ret.put("isStrongBoxBacked", false)
            
            val keyStore = KeyStore.getInstance(KEY_STORE_NAME)
            keyStore.load(null)
            keyStore.deleteEntry(alias)
            
            call.resolve(ret)
        } catch (e: Exception) {
            val ret = JSObject()
            ret.put("insideSecureHardware", true)
            ret.put("isStrongBoxBacked", false)
            ret.put("error", e.message)
            call.resolve(ret)
        }
    }
}

